pkg/controllers/subnet/subnet_controller.go

```diff
@@ -0,0 +1,271 @@
+		log.Error(err, "unable to fetch Subnet CR", "req", req.NamespacedName)
+		return ResultNormal, client.IgnoreNotFound(err)
+	}
+
+	vpcList := &v1alpha1.VPCList{}
+	if err := r.Client.List(context.Background(), vpcList, client.InNamespace(obj.Namespace)); err != nil {
+		log.Error(err, fmt.Sprintf("failed to get VPC under namespace: %s.\n", obj.Namespace))
+		return ResultRequeue, err
+	}
+	// Only one VPC under each namespace.
+	parameters := r.Service.GetSubnetParamFromPath(vpcList.Items[0].Status.NSXResourcePath)
+
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateTotal, MetricResTypeSubnet)
+		if !controllerutil.ContainsFinalizer(obj, servicecommon.FinalizerName) {
+			controllerutil.AddFinalizer(obj, servicecommon.FinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "add finalizer", "subnet", req.NamespacedName)
+				updateFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("added finalizer on subnet CR", "subnet", req.NamespacedName)
+		}
+		if err := r.Service.CreateOrUpdateSubnet(obj, parameters.OrgID, parameters.ProjectID, parameters.VPCID); err != nil {
+			log.Error(err, "operate failed, would retry exponentially", "subnet", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return ResultRequeue, err
+		}
+		if err := r.updateSubnetStatus(obj); err != nil {
+			log.Error(err, "update subnet status failed, would retry exponentially", "subnet", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return ResultRequeue, err
+		}
+		updateSuccess(r, &ctx, obj)
+	} else {
+		if controllerutil.ContainsFinalizer(obj, servicecommon.FinalizerName) {
+			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, MetricResTypeSubnet)
+			if err := r.Service.DeleteSubnet(obj.UID, parameters.OrgID, parameters.ProjectID, parameters.VPCID); err != nil {
+				log.Error(err, "deletion failed, would retry exponentially", "subnet", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			controllerutil.RemoveFinalizer(obj, servicecommon.FinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "deletion failed, would retry exponentially", "subnet", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("removed finalizer", "subnet", req.NamespacedName)
+			deleteSuccess(r, &ctx, obj)
+		} else {
+			log.Info("finalizers cannot be recognized", "subnet", req.NamespacedName)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *SubnetReconciler) updateSubnetStatus(obj *v1alpha1.Subnet) error {
+	nsxSubnets := r.Service.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetCRUID, string(obj.UID))
+	if len(nsxSubnets) == 0 {
+		return errors.New("failed to get subnet from store")
+	}
+	obj.Status.IPAddresses = obj.Status.IPAddresses[:0]
+	if len(obj.Spec.IPAddresses) > 0 {
```

Comment:
do we know if the specified IP_address Subnet is realized successfully or not?

Reply:
We don't check currently, I'll try check the api and try.

Reply:
> do we know if the specified IP_address Subnet is realized successfully or not?

In latest change, we've added logic to check realize status of Subnet.